src/udp_stream.cpp

```diff
@@ -22,22 +22,33 @@ bool raw_stream_init(raw_stream_data *sdata) {
 	sdata->dest_addr.sin_port = htons(sdata->dest_port);
 
 	if (inet_aton(sdata->dest_ip, &sdata->dest_addr.sin_addr) == 0) {
-		log(LOG_ERR, "raw_stream: inet_aton: invalid IPv4 address \"%s\"\n", sdata->dest_ip);
+		log(LOG_ERR, "udp_stream: inet_aton: invalid IPv4 address \"%s\"\n", sdata->dest_ip);
 		return false;
 	}
 
-	log(LOG_INFO, "raw_stream: sending 1 channel of 32-bit float at %d Hz to %s:%d\n", WAVE_RATE, sdata->dest_ip, sdata->dest_port);
+	log(LOG_INFO, "udp_stream: sending %s 32-bit float at %d Hz to %s:%d\n", mode==MM_MONO ? "Mono" : "Stereo", WAVE_RATE, sdata->dest_ip, sdata->dest_port);
 	return true;
 }
 
-void raw_stream_write(raw_stream_data *sdata, float *data, size_t len) {
+void udp_stream_write(udp_stream_data *sdata, float *data, size_t len) {
 	if (sdata->send_socket != -1) {
 		// Send without blocking or checking for success
 		sendto(sdata->send_socket, data, len, MSG_DONTWAIT | MSG_NOSIGNAL, (struct sockaddr *)&(sdata->dest_addr), sdata->dest_addr_len);
 	}
 }
 
-void raw_stream_shutdown(raw_stream_data *sdata) {
+void udp_stream_write(udp_stream_data *sdata, float *data_left, float *data_right, size_t len) {
+	if (sdata->send_socket != -1) {
+		float *interleaved = (float *)XCALLOC(len * 2, sizeof(float));
```

Comment:
This looks like a memory leak.

Also, the size of this buffer is known in advance, since the value of `len` is always equal to `WAVE_BATCH * sizeof(float)`. It is therefore better to allocate the buffer in `udp_stream_init()` rather than to allocate and deallocate it on every wave batch, which incurs unnecessary overhead.

Reply:
Yes it is a leak 😄 

I wrote the `udp_stream_write()` to match the parameters in `pulse_write_stream()`.  In practice it is always called with `(size_t)WAVE_BATCH * sizeof(float)` but left it as a parameter to match.

Would you rather:

1. remove `len` as a parameter, pre-allocate a buffer, and assume everything is OK
2. keep `len` as a parameter, pre-allocate a buffer, and check/assert that `len` is <= buffer's size
3. fix the memory leak by deleting the memory at the end of the function